cmd/snap-update-ns/utils.go

```diff
@@ -346,108 +318,103 @@ func splitIntoSegments(name string) ([]string, error) {
 // to request that ownership is not changed.
-func (sec *Secure) MkdirAll(name string, perm os.FileMode, uid sys.UserID, gid sys.GroupID) error {
-	logger.Debugf("secure-mkdir-all %q %v %d %d", name, perm, uid, gid)
-
+func (sec *Secure) MkdirAll(path string, perm os.FileMode, uid sys.UserID, gid sys.GroupID) error {
+	if path != filepath.Clean(path) {
+		// TODO: Remove this, it is only to make tests happy
+		return fmt.Errorf("cannot split unclean path %q", path)
+	}
 	// Only support absolute paths to avoid bugs in snap-confine when
 	// called from anywhere.
-	if !filepath.IsAbs(name) {
-		return fmt.Errorf("cannot create directory with relative path: %q", name)
-	}
-
-	// Split the path into segments.
-	segments, err := splitIntoSegments(name)
-	if err != nil {
-		return err
+	if !filepath.IsAbs(path) {
+		return fmt.Errorf("cannot create directory with relative path: %q", path)
 	}
+	base, name := filepath.Split(path)
+	base = filepath.Clean(base) // Needed to chomp the trailing slash.
 
 	// Create the prefix.
-	fd, err := sec.MkPrefix(segments, perm, uid, gid)
+	dirFd, err := sec.MkPrefix(base, perm, uid, gid)
 	if err != nil {
 		return err
 	}
-	defer sysClose(fd)
+	defer sysClose(dirFd)
 
-	if len(segments) > 0 {
-		// Create the final segment as a directory.
-		fd, err = sec.MkDir(fd, segments, len(segments)-1, perm, uid, gid)
+	if name != "" {
+		// Create the leaf as a directory.
+		leafFd, err := sec.MkDir(dirFd, base, name, perm, uid, gid)
 		if err != nil {
 			return err
 		}
-		defer sysClose(fd)
+		defer sysClose(leafFd)
 	}
 
 	return nil
 }
 
 // MkfileAll is a secure implementation of "mkdir -p $(dirname $1) && touch $1".
 //
-// This function is like secureMkdirAll but it creates an empty file instead of
+// This function is like MkdirAll but it creates an empty file instead of
 // a directory for the final path component. Each created directory component
 // is chowned to the desired user and group.
-func (sec *Secure) MkfileAll(name string, perm os.FileMode, uid sys.UserID, gid sys.GroupID) error {
-	logger.Debugf("secure-mkfile-all %q %q %d %d", name, perm, uid, gid)
-
+func (sec *Secure) MkfileAll(path string, perm os.FileMode, uid sys.UserID, gid sys.GroupID) error {
+	if path != filepath.Clean(path) {
+		// TODO: Remove this, it is only to make tests happy
+		return fmt.Errorf("cannot split unclean path %q", path)
+	}
 	// Only support absolute paths to avoid bugs in snap-confine when
 	// called from anywhere.
-	if !filepath.IsAbs(name) {
-		return fmt.Errorf("cannot create file with relative path: %q", name)
+	if !filepath.IsAbs(path) {
+		return fmt.Errorf("cannot create file with relative path: %q", path)
 	}
 	// Only support file names, not directory names.
-	if strings.HasSuffix(name, "/") {
-		return fmt.Errorf("cannot create non-file path: %q", name)
-	}
-
-	// Split the path into segments.
-	segments, err := splitIntoSegments(name)
-	if err != nil {
-		return err
+	if strings.HasSuffix(path, "/") {
```

Comment:
IIRC `filepath.Clean()` already removes the trailing slash, so this check is redundant/misplaced.

Reply:
Good point, I think this can be removed (same below)

Reply:
Actually, it cannot be cause the tests that create `/`